### PR TITLE
syncer: add prometheus metrics for syncer

### PR DESCRIPTION
**Pull request by andelf**
changes:

- add metrics
- remove old metrics atomic val
- fix typo
- fix bad log

### CODE REVIEW

File: syncer/config.go [hunk 1]
```diff
@@ -32,24 +32,22 @@ func NewConfig() *Config {
 	fs.IntVar(&cfg.ServerID, "server-id", 101, "MySQL slave server ID")
 	fs.IntVar(&cfg.WorkerCount, "c", 16, "parallel worker count")
 	fs.IntVar(&cfg.Batch, "b", 10, "batch commit count")
+	fs.StringVar(&cfg.MetricsAddr, "metrics-addr", ":2333", "metrics addr")
```
Comment (GregoryIan) [inline syncer/config.go]: 2333？ if metricsAddr is empty, just disable metrics feature like `PprofAddr`

File: syncer/db.go [hunk 2]
```diff
@@ -181,9 +181,9 @@ func genHashKey(key string) uint32 {
 	return crc32.ChecksumIEEE([]byte(key))
 }
 
-func genKeyList(columns []*column, datas []interface{}) string {
-	values := make([]string, 0, len(datas))
-	for i, data := range datas {
+func genKeyList(columns []*column, data []interface{}) string {
+	values := make([]string, 0, len(data))
+	for i, data := range data {
```
Comment (GregoryIan) [inline syncer/db.go]: `for i, datum := range data`

File: syncer/db.go [hunk 3]
```diff
@@ -198,34 +198,34 @@ func genColumnPlaceholders(length int) string {
...
-	keys := make([]string, 0, len(datas))
-	values := make([][]interface{}, 0, len(datas))
+func genInsertSQLs(schema string, table string, data [][]interface{}, columns []*column, indexColumns []*column) ([]string, []string, [][]interface{}, error) {
+	sqls := make([]string, 0, len(data))
+	keys := make([]string, 0, len(data))
+	values := make([][]interface{}, 0, len(data))
 	columnList := genColumnList(columns)
 	columnPlaceholders := genColumnPlaceholders(len(columns))
-	for _, data := range datas {
+	for _, data := range data {
```
Comment (GregoryIan) [inline syncer/db.go]: ditto

File: syncer/db.go [hunk 4]
```diff
@@ -323,13 +323,13 @@ func genUpdateSQLs(schema string, table string, datas [][]interface{}, columns [
...
-func genDeleteSQLs(schema string, table string, datas [][]interface{}, columns []*column, indexColumns []*column) ([]string, []string, [][]interface{}, error) {
-	sqls := make([]string, 0, len(datas))
-	keys := make([]string, 0, len(datas))
-	values := make([][]interface{}, 0, len(datas))
-	for _, data := range datas {
+func genDeleteSQLs(schema string, table string, data [][]interface{}, columns []*column, indexColumns []*column) ([]string, []string, [][]interface{}, error) {
+	sqls := make([]string, 0, len(data))
+	keys := make([]string, 0, len(data))
+	values := make([][]interface{}, 0, len(data))
+	for _, data := range data {
```
Comment (GregoryIan) [inline syncer/db.go]: ditto

File: syncer/db.go [hunk 5]
```diff
@@ -471,11 +471,11 @@ func parserDDLTableName(sql string) (TableName, error) {
 		res = genTableName(v.Table.Schema.L, v.Table.Name.L)
 	case *ast.DropTableStmt:
 		if len(v.Tables) != 1 {
-			return res, errors.Errorf("may resovle DDL sql failed")
+			return res, errors.Errorf("drop table with multiple tables, may resovle ddl sql failed")
 		}
 		res = genTableName(v.Tables[0].Schema.L, v.Tables[0].Name.L)
 	default:
-		return res, errors.Errorf("unkown DDL type")
+		return res, errors.Errorf("unkown ddl type %v", stmt)
```
Comment (GregoryIan) [inline syncer/db.go]: print it's type directly?

File: syncer/meta.go [hunk 6]
```diff
@@ -129,5 +129,5 @@ func (lm *LocalMeta) Check() bool {
 
 func (lm *LocalMeta) String() string {
 	pos := lm.Pos()
-	return fmt.Sprintf("binlog name = %s, binlog pos = %d", pos.Name, pos.Pos)
+	return fmt.Sprintf("binlog-name = %s, binlog-pos = %d", pos.Name, pos.Pos)
```
Comment (GregoryIan) [inline syncer/meta.go]: nice guy

File: syncer/db.go (hunk 5 above)
Comment (andelf) [inline syncer/db.go]: provide more information when an error occurs.

File: syncer/db.go (hunk 2 above)
Comment (andelf) [inline syncer/db.go]: both data are collections. so I switch to make outer data a `dataSeq`

File: syncer/metrics.go [hunk 7]
```diff
@@ -0,0 +1,73 @@
...
+	prometheus.MustRegister(binlogEventsTotal)
+	prometheus.MustRegister(binlogSkippedEventsTotal)
+	prometheus.MustRegister(sqlJobsTotal)
+	prometheus.MustRegister(sqlRetriesTotal)
+	prometheus.MustRegister(binlogMetaPos)
+
+	// Expose the registered metrics via HTTP.
+	go func() {
+		http.Handle("/metrics", promhttp.Handler())
+		log.Fatal(http.ListenAndServe(addr, nil))
```
Comment (nolouch) [inline syncer/metrics.go]: you can share one address with pprof. no need listen twice.

**Comment by GregoryIan** [comment]
LGTM

**Comment by nolouch** [comment]
LGTM

**Pull request merged**